backend/src/test/java/com/google/backend/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,142 @@
+    new LocalDatastoreServiceTestConfig(), new LocalUserServiceTestConfig());
+
+  private static final String CONTENT_TYPE = "text/html;";
+  private static final String TEST_EMAIL = "test@example.com";
+  private static final String REDIRECT_URL = "/";
+  private static final String LOGIN_STATUS = "logged in";
+  private static final String LOGOUT_STATUS = "stranger";
+  private static final String ENTITY_KIND = "User";
+  private static final int MAX_QUERIES = 1000;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGet_GetsCorrectResponseLoggedIn() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain("localhost");
+    UserService userService = UserServiceFactory.getUserService();
+    String logoutUrl = userService.createLogoutURL(REDIRECT_URL);
+    String loginResponse = LOGIN_STATUS + "\n" + logoutUrl;
+    
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    loginServlet.doGet(request, response);
+
+    verify(response).setContentType(CONTENT_TYPE);
+    assertEquals(stringWriter.getBuffer().toString().trim(), loginResponse);
+  }
+
+  @Test
+  public void doGet_UserDoesNotExistCreateNewEntity() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain("localhost");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    loginServlet.doGet(request, response);
+    
+    Query query =
+          new Query(ENTITY_KIND).setFilter(
+            new Query.FilterPredicate("email", Query.FilterOperator.EQUAL, TEST_EMAIL));
+    Entity result = datastore.prepare(query).asSingleEntity();
+    assertNotNull(result);
+  }
+
+  @Test
+  public void doGet_UserExistsNoNewEntity() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain("localhost");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity newUser = new Entity(ENTITY_KIND);
+    newUser.setProperty("email", TEST_EMAIL);
+    datastore.put(newUser);
+    
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    loginServlet.doGet(request, response);
+
+    Query query =
+          new Query(ENTITY_KIND).setFilter(
+            new Query.FilterPredicate("email", Query.FilterOperator.EQUAL, TEST_EMAIL));
+    int numComments = datastore.prepare(query).countEntities(
```

Comment:
ooc, why do we have `numComments` here?

i thought we're just storing users in the datastore?

Reply:
i forgot to rename my variable O.O sorry

Reply:
lol no worries!